prod-profile configuration pollutes test profile in integration tests

### Describe the bug

The bug was introduces in Quarkus version `3.5.2`; it was not present in version `3.5.1`.

When we have configuration values for the `prod` profile in the `application.properties`, they "pollute" the profile set for `quarkus.test.integration-test-profile`, possibly leading to the test failing because the application fails to start.

An example:

We use an application with a postgres database, `quarkus-smallrye-health` and the following `application.properties`:

```properties
quarkus.http.test-port=-1
quarkus.test.integration-test-profile=test
quarkus.datasource.db-kind=postgresql
%prod.quarkus.datasource.username=${DB_USER}
%prod.quarkus.datasource.password=${DB_PASSWORD}
%prod.quarkus.datasource.jdbc.additional-jdbc-properties.foo=${FOO}
```

We test that the database connection (provided by a devservice) is established with the following `@QuarkusTest`:

```java
package de.turing85.quarkus.properties;

import io.quarkus.test.junit.QuarkusTest;
import io.restassured.RestAssured;
import org.junit.jupiter.api.Test;

import static org.hamcrest.CoreMatchers.is;

@QuarkusTest
class HealthTest {
  @Test
  void healthTest() {
    // @formatter:off
    RestAssured
        .when().get("/q/health")
        .then()
            .body("checks.find { it.name = 'Database connections health check' }.status", is("UP"))
            .body("checks.find { it.name = 'Database connections health check' }.data.'<default>'", is("UP"));
    // @formatter:on
  }
}
```

and the following `@QuarkusIntegrationTest`:

```java
package de.turing85.quarkus.properties;

import io.quarkus.test.junit.QuarkusIntegrationTest;

@QuarkusIntegrationTest
class HealthTestIT extends HealthTest {
}
```

When we now run the tests, the application fails to start in the failsafe tests (the surefire tests work fine):

```bash
$ ./mvnw clean verify
...
[INFO] [stdout] ERROR: Failed to start application (with profile [test])
[INFO] [stdout] java.lang.RuntimeException: Failed to start quarkus
[INFO] [stdout] 	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
[INFO] [stdout] 	at io.quarkus.runtime.Application.start(Application.java:101)
[INFO] [stdout] 	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:111)
[INFO] [stdout] 	at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
[INFO] [stdout] 	at io.quarkus.runtime.Quarkus.run(Quarkus.java:44)
[INFO] [stdout] 	at io.quarkus.runtime.Quarkus.run(Quarkus.java:124)
[INFO] [stdout] 	at io.quarkus.runner.GeneratedMain.main(Unknown Source)
[INFO] [stdout] 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[INFO] [stdout] 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
[INFO] [stdout] 	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
[INFO] [stdout] 	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
[INFO] [stdout] Caused by: java.util.NoSuchElementException: SRCFG00011: Could not expand value FOO in property quarkus.datasource.jdbc.additional-jdbc-properties.foo
[INFO] [stdout] 	at io.smallrye.config.ExpressionConfigSourceInterceptor$1.accept(ExpressionConfigSourceInterceptor.java:85)
[INFO] [stdout] 	at io.smallrye.config.ExpressionConfigSourceInterceptor$1.accept(ExpressionConfigSourceInterceptor.java:61)
[INFO] [stdout] 	at io.smallrye.common.expression.ExpressionNode.emit(ExpressionNode.java:22)
[INFO] [stdout] 	at io.smallrye.common.expression.Expression.evaluateException(Expression.java:56)
[INFO] [stdout] 	at io.smallrye.common.expression.Expression.evaluate(Expression.java:70)
[INFO] [stdout] 	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:61)
[INFO] [stdout] 	at io.smallrye.config.ExpressionConfigSourceInterceptor.getValue(ExpressionConfigSourceInterceptor.java:35)
[INFO] [stdout] 	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
[INFO] [stdout] 	at io.smallrye.config.SecretKeysHandlerConfigSourceInterceptor.getValue(SecretKeysHandlerConfigSourceInterceptor.java:25)
[INFO] [stdout] 	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
[INFO] [stdout] 	at io.smallrye.config.FallbackConfigSourceInterceptor.getValue(FallbackConfigSourceInterceptor.java:24)
[INFO] [stdout] 	at io.smallrye.config.SmallRyeConfigSourceInterceptorContext.proceed(SmallRyeConfigSourceInterceptorContext.java:20)
[INFO] [stdout] 	at io.smallrye.config.SmallRyeConfig.getConfigValue(SmallRyeConfig.java:322)
[INFO] [stdout] 	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:229)
[INFO] [stdout] 	at io.smallrye.config.ConfigMappingProvider.lambda$processLazyMapValue$0(ConfigMappingProvider.java:504)
[INFO] [stdout] 	at io.smallrye.config.ConfigMappingProvider.mapConfigurationInternal(ConfigMappingProvider.java:990)
[INFO] [stdout] 	at io.smallrye.config.ConfigMappingProvider.lambda$mapConfiguration$3(ConfigMappingProvider.java:962)
[INFO] [stdout] 	at io.smallrye.config.SecretKeys.doUnlocked(SecretKeys.java:28)
[INFO] [stdout] 	at io.smallrye.config.ConfigMappingProvider.mapConfiguration(ConfigMappingProvider.java:962)
[INFO] [stdout] 	at io.smallrye.config.ConfigMappings.mapConfiguration(ConfigMappings.java:91)
[INFO] [stdout] 	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:699)
[INFO] [stdout] 	at io.quarkus.runtime.generated.Config.readConfig(Unknown Source)
[INFO] [stdout] 	at io.quarkus.runtime.generated.Config.createRunTimeConfig(Unknown Source)
[INFO] [stdout] 	at io.quarkus.deployment.steps.RuntimeConfigSetup.deploy(Unknown Source)
[INFO] [stdout] 	... 11 more
...
```

### Expected behavior

The tests should succeed.

### Actual behavior

The application fails to start with the above error

### How to Reproduce?

Reproducer:
- clone https://github.com/turing85/quarkus-properties:
  ```bash
  git clone https://github.com/turing85/quarkus-properties.git && quarkus-properties
  ```
- build the application, run the tests:
  ```bash
  ./mvnw clean verify
  ```
- Observe the error mentioned above
- Open `pom.xml`, set `quarkus.platform.version` from `3.5.2` to `3.5.1`
- Re-run the tests:
  ```bash
  ./mvnw clean verify
  ```
- Observe that the build succeeds

### Output of `uname -a` or `ver`
```
Linux XXX 5.15.0-52-generic #58-Ubuntu SMP Thu Oct 13 08:03:55 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

### Output of `java -version`
```
openjdk version "21.0.2" 2024-01-16 LTS
OpenJDK Runtime Environment Temurin-21.0.2+13 (build 21.0.2+13-LTS) 
OpenJDK 64-Bit Server VM Temurin-21.0.2+13 (build 21.0.2+13-LTS, mixed mode, sharing)
```
### Quarkus version or git rev

`3.5.2` onwards

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
```
Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae) 
Maven home: /home/marco/.m2/wrapper/dists/apache-maven-3.9.6-bin/3311e1d4/apache-maven-3.9.6
Java version: 21.0.2, vendor: Eclipse Adoptium, runtime: /opt/java/mandrel/23.1.2.0-java21
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.15.0-52-generic", arch: "amd64", family: "unix"
```
### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java
@@ -78,6 +78,7 @@ import io.smallrye.config.Converters;
 import io.smallrye.config.EnvConfigSource;
 import io.smallrye.config.KeyMap;
 import io.smallrye.config.KeyMapBackedConfigSource;
+import io.smallrye.config.ProfileConfigSourceInterceptor;
 import io.smallrye.config.PropertiesConfigSource;
 import io.smallrye.config.SecretKeys;
 import io.smallrye.config.SmallRyeConfig;
@@ -640,7 +641,7 @@ public final class BuildTimeConfigurationReader {
                     unknownBuildProperties.remove(property);
                     ConfigValue value = config.getConfigValue(property);
                     if (value != null && value.getRawValue() != null) {
-                        allBuildTimeValues.put(property, value.getRawValue());
+                        allBuildTimeValues.put(value.getNameProfiled(), value.getRawValue());
                     }
                 }
             }
@@ -652,8 +653,8 @@ public final class BuildTimeConfigurationReader {
                     unknownBuildProperties.remove(property);
                     ConfigValue value = config.getConfigValue(property);
                     if (value != null && value.getRawValue() != null) {
-                        allBuildTimeValues.put(property, value.getRawValue());
-                        buildTimeRunTimeValues.put(property, value.getRawValue());
+                        allBuildTimeValues.put(value.getNameProfiled(), value.getRawValue());
+                        buildTimeRunTimeValues.put(value.getNameProfiled(), value.getRawValue());
                     }
                 }
             }
@@ -665,7 +666,7 @@ public final class BuildTimeConfigurationReader {
                     unknownBuildProperties.remove(property);
                     ConfigValue value = config.getConfigValue(property);
                     if (value != null && value.getRawValue() != null) {
-                        runTimeValues.put(property, value.getRawValue());
+                        runTimeValues.put(value.getNameProfiled(), value.getRawValue());
                     }
                 }
             }
@@ -1073,8 +1074,7 @@ public final class BuildTimeConfigurationReader {
 
             Set<String> properties = new HashSet<>();
 
-            // We build a new Config to also apply the interceptor chain, this includes the active profile. A property
-            // may only exist in its profiled name format, but it requires recording in the unprofiled name
+            // We build a new Config to also apply the interceptor chain to generate any additional properties.
             SmallRyeConfigBuilder builder = ConfigUtils.emptyConfigBuilder();
             builder.getSources().clear();
             builder.getSourceProviders().clear();
@@ -1108,8 +1108,10 @@ public final class BuildTimeConfigurationReader {
                             return Collections.emptySet();
                         }
                     });
+
+            String[] profiles = config.getProfiles().toArray(new String[] {});
             for (String property : builder.build().getPropertyNames()) {
-                properties.add(property);
+                properties.add(ProfileConfigSourceInterceptor.activeName(property, profiles));
             }
 
             return properties;

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/BuildTimeConfigurationReader.java
@@ -1109,7 +1109,7 @@ public final class BuildTimeConfigurationReader {
                         }
                     });
 
-            String[] profiles = config.getProfiles().toArray(new String[] {});
+            String[] profiles = config.getProfiles().toArray(String[]::new);
             for (String property : builder.build().getPropertyNames()) {
                 properties.add(ProfileConfigSourceInterceptor.activeName(property, profiles));
             }

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
@@ -92,6 +92,7 @@ import io.smallrye.config.ConfigSourceFactory;
 import io.smallrye.config.ConfigSourceInterceptor;
 import io.smallrye.config.ConfigSourceInterceptorFactory;
 import io.smallrye.config.DefaultValuesConfigSource;
+import io.smallrye.config.ProfileConfigSourceInterceptor;
 import io.smallrye.config.SecretKeysHandler;
 import io.smallrye.config.SecretKeysHandlerFactory;
 import io.smallrye.config.SmallRyeConfig;
@@ -217,6 +218,14 @@ public class ConfigGenerationBuildStep {
             defaultValues.put(e.getKey(), e.getValue());
         }
         // Recorded values from build time from any other source (higher ordinal then defaults, so override)
+        String[] profiles = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class).getProfiles().toArray(new String[] {});
+        for (Map.Entry<String, String> entry : configItem.getReadResult().getRunTimeValues().entrySet()) {
+            // Runtime values may contain active profiled names that override sames names in defaults
+            // We need to keep the original name definition in case a different profile is used to run the app
+            String activeName = ProfileConfigSourceInterceptor.activeName(entry.getKey(), profiles);
+            defaultValues.remove(activeName);
+            defaultValues.put(entry.getKey(), entry.getValue());
+        }
         defaultValues.putAll(configItem.getReadResult().getRunTimeValues());
 
         Set<String> converters = discoverService(Converter.class, reflectiveClass);

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigGenerationBuildStep.java
@@ -218,7 +218,7 @@ public class ConfigGenerationBuildStep {
             defaultValues.put(e.getKey(), e.getValue());
         }
         // Recorded values from build time from any other source (higher ordinal then defaults, so override)
-        String[] profiles = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class).getProfiles().toArray(new String[] {});
+        String[] profiles = ConfigProvider.getConfig().unwrap(SmallRyeConfig.class).getProfiles().toArray(String[]::new);
         for (Map.Entry<String, String> entry : configItem.getReadResult().getRunTimeValues().entrySet()) {
             // Runtime values may contain active profiled names that override sames names in defaults
             // We need to keep the original name definition in case a different profile is used to run the app

--- a/integration-tests/test-extension/extension/deployment/src/main/java/io/quarkus/extest/deployment/BuildTimeCustomConfigBuilder.java
+++ b/integration-tests/test-extension/extension/deployment/src/main/java/io/quarkus/extest/deployment/BuildTimeCustomConfigBuilder.java
@@ -14,6 +14,9 @@ public class BuildTimeCustomConfigBuilder implements SmallRyeConfigBuilderCustom
                 new PropertiesConfigSource(Map.of("prop.recorded.from.btconfigsource", "1234"), "BuildTimeConfigSource", 100));
         // for RecorderRuntimeConfigTest
         builder.withSources(
-                new PropertiesConfigSource(Map.of("recorded.property", "from-application"), "BuildTimeConfigSource", 250));
+                new PropertiesConfigSource(Map.of(
+                        "recorded.property", "from-application",
+                        "%test.recorded.profiled.property", "from-application",
+                        "%test.quarkus.mapping.rt.record-profiled", "from-application"), "BuildTimeConfigSource", 250));
     }
 }

--- a/integration-tests/test-extension/extension/deployment/src/main/java/io/quarkus/extest/deployment/TestProcessor.java
+++ b/integration-tests/test-extension/extension/deployment/src/main/java/io/quarkus/extest/deployment/TestProcessor.java
@@ -490,6 +490,10 @@ public final class TestProcessor {
     @BuildStep
     void recordProperty(BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfigurationDefault) {
         runTimeConfigurationDefault.produce(new RunTimeConfigurationDefaultBuildItem("recorded.property", "from-build-step"));
+        runTimeConfigurationDefault
+                .produce(new RunTimeConfigurationDefaultBuildItem("recorded.profiled.property", "should-not-be-recorded"));
+        runTimeConfigurationDefault.produce(
+                new RunTimeConfigurationDefaultBuildItem("quarkus.mapping.rt.record-profiled", "should-not-be-recorded"));
     }
 
     public static final class Never implements BooleanSupplier {

--- a/integration-tests/test-extension/extension/deployment/src/test/java/io/quarkus/config/RecorderRuntimeConfigTest.java
+++ b/integration-tests/test-extension/extension/deployment/src/test/java/io/quarkus/config/RecorderRuntimeConfigTest.java
@@ -1,15 +1,23 @@
 package io.quarkus.config;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.Optional;
 
 import jakarta.inject.Inject;
 
+import org.eclipse.microprofile.config.spi.ConfigSource;
 import org.jboss.shrinkwrap.api.ShrinkWrap;
 import org.jboss.shrinkwrap.api.spec.JavaArchive;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
+import io.quarkus.extest.runtime.config.TestMappingRunTime;
 import io.quarkus.test.QuarkusUnitTest;
+import io.smallrye.config.DefaultValuesConfigSource;
 import io.smallrye.config.SmallRyeConfig;
 
 public class RecorderRuntimeConfigTest {
@@ -19,9 +27,24 @@ public class RecorderRuntimeConfigTest {
 
     @Inject
     SmallRyeConfig config;
+    @Inject
+    TestMappingRunTime mappingRunTime;
 
     @Test
     void runtimeConfig() {
         assertEquals("from-application", config.getRawValue("recorded.property"));
+        assertEquals("from-application", config.getRawValue("recorded.profiled.property"));
+        assertEquals("from-application", config.getRawValue("quarkus.mapping.rt.record-profiled"));
+        assertTrue(mappingRunTime.recordProfiled().isPresent());
+        assertEquals("from-application", mappingRunTime.recordProfiled().get());
+
+        // Make sure that when we merge all the defaults, we override the non-profiled name
+        Optional<ConfigSource> configSource = config.getConfigSource("DefaultValuesConfigSource");
+        assertTrue(configSource.isPresent());
+        DefaultValuesConfigSource defaultValuesConfigSource = (DefaultValuesConfigSource) configSource.get();
+        assertNotNull(defaultValuesConfigSource.getValue("%test.recorded.profiled.property"));
+        assertNull(defaultValuesConfigSource.getValue("recorded.profiled.property"));
+        assertNotNull(defaultValuesConfigSource.getValue("%test.quarkus.mapping.rt.record-profiled"));
+        assertNull(defaultValuesConfigSource.getValue("quarkus.mapping.rt.record-profiled"));
     }
 }

--- a/integration-tests/test-extension/extension/runtime/src/main/java/io/quarkus/extest/runtime/config/TestMappingRunTime.java
+++ b/integration-tests/test-extension/extension/runtime/src/main/java/io/quarkus/extest/runtime/config/TestMappingRunTime.java
@@ -22,6 +22,9 @@ public interface TestMappingRunTime {
     /** Record values from env test **/
     Optional<String> record();
 
+    /** Record values with named profile **/
+    Optional<String> recordProfiled();
+
     interface Group {
         /**
          * A Group value.

